Fix for DATENAME() gives incorrect value with TZOFFSET part

manisha-deshpande opened the pull request:
### Description
Currently, DATENAME function returns a numeric value for TZOFFSET datepart, which does not match SQL Server's text output.

**Root cause**: The function definition of sys.datename explicitly calls the DATEPART function for most argument values, including TZOFFSET, thereby returning a numerical offset value instead of the required text format (+/-HH:MM).

**Fix**: Modified the sys function's SQL definition to extract and return the timezone offset substring from the input - similar to how the offset_string is determined in DATETRUNC/DATEBUCKET - if the argument matches tzoffset. Furthermore, translated the function from SQL language to plpgsql in order to enable raising an exception for unsupported input datetime datatypes such as date, time, datetime and smalldatetime.

**Additional changes**: Updated an existing jdbc test case `datepart-vu-verify.sql` to avoid 'failing' table insertions of `DATENAME(TZOFFSET, cursor(date))` for incompatible date types: datetime, smalldatetime, datetimeoffset.

`IF NOT (@datepart = 'tzoffset' AND @datatype NOT IN ('DATETIMEOFFSET', 'DATETIME2'))
`
Added corresponding version upgrade tests.

### Issues Resolved

BABEL-5846

### Test Scenarios Covered ###
* **Use case based -**
```
-- Test positive/negative offsets
SELECT DATENAME(TZOFFSET, '2025-06-03 14:30:15 +01:59');
GO
~~START~~
text
+01:59
~~END~~

select datename(TZoffset, CAST('2016-12-26 23:30:05.523456+8'AS datetimeoffset));
GO
~~START~~
text
+08:00
~~END~~

SELECT DATENAME(TZOFFSET, CAST('2025-06-03 14:30:15 +01:59' AS DATETIMEOFFSET));
GO
~~START~~
text
+01:59
~~END~~

SELECT DATENAME(TZOFFSET, CAST('2025-06-03 14:30:15 -01:59' AS DATETIMEOFFSET));
GO
~~START~~
text
-01:59
~~END~~
```

* **Boundary conditions -**
```
-- Test case for NULL input
SELECT DATENAME(TZOFFSET, NULL);
GO
~~START~~
text
<NULL>
~~END~~

-- Test case for empty string
SELECT DATENAME(TZOFFSET, '');
GO
~~START~~
text
+00:00
~~END~~

-- Boundary values
SELECT DATENAME(TZOFFSET, CAST('2025-06-03 14:30:15 +14:00' AS DATETIMEOFFSET));
GO
~~START~~
text
+14:00
~~END~~

SELECT DATENAME(TZOFFSET, CAST('2025-06-03 14:30:15 -14:00' AS DATETIMEOFFSET));
GO
~~START~~
text
-14:00
~~END~~

-- Supported date time datatype inputs
SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15' AS DATETIME2)) AS datetime2_type;
GO
~~START~~
text
+00:00
~~END~~

SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15' AS DATETIMEOFFSET)) AS datetimeoffset_type;
GO
~~START~~
text
+00:00
~~END~~
```

* **Arbitrary inputs -**
```
    -- Invalid date time inputs
SELECT DATENAME(TZOFFSET, '2025-06-0314:30:15+01:59');
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: Conversion failed when converting date and/or time from character string.)~~

-- Test date inputs without offset specified
SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15.1234567' AS DATETIMEOFFSET));
GO
~~START~~
text
+00:00
~~END~~

SELECT datename(TZOFFSET, CAST('2025-06-03' AS DATETIMEOFFSET));
GO
~~START~~
text
+00:00
~~END~~

SELECT datename(TZOFFSET, CAST('14:30:15.1234567' AS DATETIMEOFFSET));
GO
~~START~~
text
+00:00
~~END~~

SELECT datename(TZOFFSET, CAST('' AS DATETIMEOFFSET));
GO
~~START~~
text
+00:00
~~END~~
```

* **Negative test cases -**
```
-- Error Scenarios
    -- Unsupported date time datatype inputs
SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15 +01:30' AS DATE)) AS date_type;
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: The datepart tzoffset is not supported by date function datename for data type date.)~~


SELECT datename(TZOFFSET, CAST('14:30:15' AS TIME)) AS time_type;
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: The datepart tzoffset is not supported by date function datename for data type time without time zone.)~~


SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15' AS DATETIME)) AS datetime_type;
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: The datepart tzoffset is not supported by date function datename for data type datetime.)~~


SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15' AS SMALLDATETIME)) AS smalldatetime_type;
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: The datepart tzoffset is not supported by date function datename for data type smalldatetime.)~~



    -- Invalid date time inputs
SELECT datename(TZOFFSET, CAST('2025-06-0314:30:15+130' AS DATETIMEOFFSET));
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: invalid input syntax for type datetimeoffset: "2025-06-0314:30:15+130")~~


SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15 -16:00' AS DATETIMEOFFSET));
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: time zone displacement out of range: "2025-06-03 14:30:15 -16:00")~~


SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15 +10:60' AS DATETIMEOFFSET));
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: time zone displacement out of range: "2025-06-03 14:30:15 +10:60")~~


SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15 +01:30' AS DATETIMEOFFSET) AT TIME ZONE 'Invalid Time Zone');
GO
~~START~~
text
~~ERROR (Code: 33557097)~~

~~ERROR (Message: Argument data type or the parameter Invalid Time Zone provided to AT TIME ZONE clause is invalid.)~~


SELECT datename(TZOFFSET, SWITCHOFFSET(CAST('2025-06-03 14:30:15 +01:30' AS DATETIMEOFFSET), 'Invalid Offset'));
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: The timezone provided to builtin function todatetimeoffset is invalid.)~~


SELECT datename(TZOFFSET, TODATETIMEOFFSET(CAST('2025-06-03 14:30:15' AS DATETIME2), 'Invalid Offset'));
GO
~~ERROR (Code: 33557097)~~

~~ERROR (Message: The timezone provided to builtin function todatetimeoffset is invalid.)~~

```

* **Integration test cases -**
```
-- Test with other functions
SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15 +01:30' AS DATETIMEOFFSET) AT TIME ZONE 'Pacific Standard Time');
GO
~~START~~
text
-07:00
~~END~~

SELECT DATENAME(TZOFFSET, SWITCHOFFSET(CAST('2025-06-03 14:30:15 +01:30' AS DATETIMEOFFSET), '-08:00'));
GO
~~START~~
text
-08:00
~~END~~

SELECT DATENAME(TZOFFSET, TODATETIMEOFFSET(CAST('2025-06-03 14:30:15' AS DATETIME2), '-03:30'));
GO
~~START~~
text
-03:30
~~END~~
```
* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

manisha-deshpande committed:
Fix for DATENAME() gives incorrect value with TZOFFSET part

Currently, DATENAME function returns a numeric value for TZOFFSET
datepart, which does not match SQL Server's text output.

Root cause: The function definition of sys.datename explicitly calls the
DATEPART function for most argument values, including TZOFFSET, thereby
returning a numerical offset value instead of the required text format
(+/-HH:MM).

Fix: Modified the sys function's SQL definition to extract and return
the timezone offset substring from the input - similar to how the
offset_string is determined in DATETRUNC/DATEBUCKET - if the argument
matches tzoffset. Furthermore, translated the function from SQL language
to plpgsql in order to enable raising an exception for unsupported input
datetime datatypes such as date, time, datetime and smalldatetime.

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande committed:
Added cast to datetimeoffset in datename function

In the datename function, before returning the tzoffset substring,
called a sys helper function `babelfish_conv_string_to_datetimeoffset` to
validate the date string and raise exceptions for invalid formats.

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande committed:
Updated related test case

DATENAME should support only  datetime2 and datetimeoffset datetime datatypes
as input when the datepart argument is TZOFFSET. SQL Server throws an
unsupported error for datatypes date, time, datetime, smalldatetime when
passed to datename(TZOFFSET, date). Modified a related test case to skip
failing inserts for the mentioned datepart-datatype combinations.

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande commented:
TODO: Will look into version upgrade test failures
Edit: DONE

manisha-deshpande commented on test/JDBC/input/functions/datepart-vu-verify.sql:
```diff
@@ -249,34 +249,37 @@ BEGIN
 
         WHILE @@FETCH_STATUS = 0
         BEGIN
-            -- Cursor for test dates
```
TODO: Fix space/tab. only added an if condition and moved some logic into that if, but diff loaded differently.

manisha-deshpande committed:
Added upgrade script and schedule file changes

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

Anikait143 commented on test/JDBC/input/datename_tzoffset.sql:
We would have to create cleanup/prepare/verify files to work on upgrade checks

Anikait143 commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2051,36 +2051,59 @@ LANGUAGE C IMMUTABLE PARALLEL SAFE;
 
 CREATE OR REPLACE FUNCTION sys.datename(IN dp PG_CATALOG.TEXT, IN arg anyelement) RETURNS TEXT AS 
```
Return type of datename function should be `NVARCHAR`

manisha-deshpande replied:
This is true, according to the official tsql documentation for datename.
But updating this now, updates all the datetime datatype tests that reference datename(). 
```
---  .../expected/babel_datetime-vu-verify.out
+++ .../output/babel_datetime-vu-verify.out
 select datename(year, cast('2002-05-23 23:41:29.998' as datetime))
 go
 ~~START~~
-text
+nvarchar
```
This would require upgrade files for all the tests below. 
```
[ERROR] Errors: 
[ERROR]   TestQueryFile.TestQueryBatch(String)[12] » Throwable babel_datetime-vu-verify ...
[ERROR]   TestQueryFile.TestQueryBatch(String)[14] » Throwable babel_datetime FAILED! Ou...
[ERROR]   TestQueryFile.TestQueryBatch(String)[15] » Throwable babel_datetime2 FAILED! O...
[ERROR]   TestQueryFile.TestQueryBatch(String)[18] » Throwable babel_function FAILED! Ou...
[ERROR]   TestQueryFile.TestQueryBatch(String)[19] » Throwable babel_smalldatetime FAILE...
```
Should this change be picked up as part of this TZOFFSET specific bug, or a separate task?

manisha-deshpande replied:
Already tracked in a separate task: BABEL-5836

Anikait143 commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2051,36 +2051,59 @@ LANGUAGE C IMMUTABLE PARALLEL SAFE;
 $BODY$
-STRICT
-LANGUAGE sql IMMUTABLE;
+LANGUAGE plpgsql IMMUTABLE;
```
Since it involves date/time operations, it's better to define function as `STABLE`

manisha-deshpande replied:
Making it _STABLE_ or retaining _STRICT_ is causing `babel_function` to fail with:
```
--- /local/babelfish_extensions/test/JDBC/./expected/babel_function.out	2025-06-20 04:47:53.280719934 +0000
+++ /local/babelfish_extensions/test/JDBC/./output/babel_function.out	2025-06-20 05:59:51.855832545 +0000
@@ -1195,6 +1195,8 @@
 -- test invalid argument, expect error
 select datename(invalid_interval, cast('2016-12-26 23:30:05.523456' as date));
 GO
+~~START~~
+text
 ~~ERROR (Code: 33557097)~~
 
 ~~ERROR (Message: 'invalid_interval' is not a recognized datepart option)~~
```
Is this expected? If so, will it require upgrade tests for `babel_function` as well?

KushaalShroff replied:
As per documentation:

The return value depends on the language environment set by using [SET LANGUAGE](https://learn.microsoft.com/en-us/sql/t-sql/statements/set-language-transact-sql?view=sql-server-ver16), and by the [Configure the default language Server Configuration Option](https://learn.microsoft.com/en-us/sql/database-engine/configure-windows/configure-the-default-language-server-configuration-option?view=sql-server-ver16) of the login. The return value depends on [SET DATEFORMAT](https://learn.microsoft.com/en-us/sql/t-sql/statements/set-dateformat-transact-sql?view=sql-server-ver16) if date is a string literal of some formats. SET DATEFORMAT does not change the return value when the date is a column expression of a date or time data type.

When the date parameter has a date data type argument, the return value depends on the setting specified by [SET DATEFIRST](https://learn.microsoft.com/en-us/sql/t-sql/statements/set-datefirst-transact-sql?view=sql-server-ver16).

The documentation sways the function to be a stable one and not immutable.
We should see if we support such configs and add tests for them

manisha-deshpande replied:
Babelfish currently supports:
- **SET LANGUAGE** us_english; => same result always ~ can remain IMMUTABLE
- **SET DATEFIRST** 1 or 7 varies the result for: `SELECT DATENAME(week, '2025-01-05');` ~ DATENAME should be rewritten as STABLE, but works correctly even without STABLE (see example below)
- **SET DATEFORMAT** (MDY, YMD only. DMY rewritten to MDY, YDM to YMD. MYD,DYM not supported)=> results for an input to DATENAME would vary for MDY-DMY oy YMD-YDM combinations but they are currently rewritten to the same format and output ~ can remain IMMUTABLE

```
1> SET DATEFIRST 7;
2> SELECT DATENAME(week, '2025-01-05');
3> go
datename                                                                                                                                                                                                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
2                                                                                                                                                                                                                                                               

(1 rows affected)
1>  SET DATEFIRST 1;
2> go
1> SELECT DATENAME(week, '2025-01-05');
2> go
datename                                                                                                                                                                                                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1                                                                                                                                                                                                                                                               

(1 rows affected)
1> select @@version;
2> go
version                                                                                                                                                                                                                                                         
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Babelfish for PostgreSQL with SQL Server Compatibility - 12.0.2000.8
Jun 26 2025 00:39:21
Copyright (c) Amazon Web Services
PostgreSQL 17.4 on x86_64-pc-linux-gnu (Babelfish 5.3.0)                                                                            

(1 rows affected)
```

**Only difference in output:**
-IMMUTABLE
+STABLE
```
 SELECT datename(TZOFFSET, CAST('2025-06-03 14:30:15' AS SMALLDATETIME)) AS smalldatetime_type;
 GO
+~~START~~
+text
 ~~ERROR (Code: 33557097)~~
 
 ~~ERROR (Message: The datepart tzoffset is not supported by date function datename for data type smalldatetime.)~~
```

KushaalShroff replied:
That is not how we test volatility and results will not be changing for simple select datename queries. Please check https://www.postgresql.org/docs/current/xfunc-volatility.html.
For safe side, I lean towards making it stable.

manisha-deshpande replied:
Added relevant test cases. 
Function should most probably be STABLE but was previously changed to immutable as part of Task: BABEL-1475 `Analyze and make date related functions immutable in babelfish`.

manisha-deshpande committed:
Changed testfile into prepare/verify/cleanup structure

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande committed:
Removed sys.datename dependency from upgrade_validation

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande committed:
Resolved failing version upgrade test

Moved datepart, datename inserts SQL from the verify file of
datepart-before-16_10-or-17_6 to the prepare file in a view and tested
the view in the verify file. This is to avoid the verify SQL from
referencing the new definition of sys.DATENAME(TZOFFSET,...) and failing
in the version upgrade tests.

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande committed:
Added config tests to check if DATENAME is STABLE

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande commented:
Updating the DATENAME function from STABLE to IMMUTABLE fails an existing test case `BABEL_1475` due to disallowing the use of _computed columns_ with DATENAME:
```
--- /home/runner/work/babelfish_extensions/babelfish_extensions/test/JDBC/./expected/BABEL-1475-vu-prepare.out	2025-06-27 21:14:12.077114693 +0000
+++ /home/runner/work/babelfish_extensions/babelfish_extensions/test/JDBC/./output/BABEL-1475-vu-prepare.out	2025-06-27 21:29:59.986827039 +0000
@@ -44,7 +44,9 @@
 CREATE TABLE BABEL_1475_vu_prepare_datename (dt date, year as DATENAME(year, dt), month as DATENAME(month, dt), weekday as DATENAME(dow, dt), dayofyear as DATENAME(dayofyear, dt), day as DATENAME(day, dt));
 INSERT INTO BABEL_1475_vu_prepare_datename (dt) values ('1912-10-25');
 GO
-~~ROW COUNT: 1~~
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: generation expression is not immutable)~~
 -- Test DATEPART function for computed columns
--- /home/runner/work/babelfish_extensions/babelfish_extensions/test/JDBC/./expected/BABEL-1475-vu-verify.out	2025-06-27 21:14:12.077114693 +0000
+++ /home/runner/work/babelfish_extensions/babelfish_extensions/test/JDBC/./output/BABEL-1475-vu-verify.out	2025-06-27 21:30:00.017827323 +0000
@@ -55,10 +55,9 @@
 -- Test DATENAME function for computed columns
 SELECT year, month, weekday, dayofyear, day from BABEL_1475_vu_prepare_datename;
 GO
-~~START~~
-text#!#text#!#text#!#text#!#text
-1912#!#October#!#Friday#!#299#!#25
-~~END~~
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: relation "babel_1475_vu_prepare_datename" does not exist)~~
```
cc: @KushaalShroff @Anikait143

manisha-deshpande committed:
Reverting STABLE to IMMUTABLE for DATENAME

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande committed:
Refactored version upgrade tests

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande committed:
Added version upgrade tests for datename_tzoffset

Added datename_tzoffset-before-16_10-or-17_6-vu-prepare/verify/cleanup
test files to represent the difference in DATENAME output format before
function update. Previously, timezone offset output was numeric, now it
is in an offset string format '+/- HH:MM'.

Task: BABEL-5846
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

KushaalShroff commented:
Please check Test failures!

Anikait143 commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -2056,36 +2056,59 @@ LANGUAGE C IMMUTABLE PARALLEL SAFE;
 $BODY$
-STRICT
-LANGUAGE sql IMMUTABLE;
+LANGUAGE plpgsql IMMUTABLE;
```
Should we mark this as `PARALLEL SAFE`?

forestkeeper replied:
It's fine not to set parallel safe in this PR.

manisha-deshpande replied:
Agreed. Furthermore, as discussed, finding:
DATENAME's internal function calls are also currently marked parallel unsafe:
```
babelfish_db=# select proname, proparallel from pg_proc where proname like 'datepart';
 proname  | proparallel 
----------+-------------
 datepart | u
 datepart | u
(2 rows)
babelfish_db=# select proname, proparallel from pg_proc where proname like 'right';
 proname | proparallel 
---------+-------------
 right   | s
 right   | s
 right   | s
 right   | s
(4 rows)
babelfish_db=# select proname, proparallel from pg_proc where proname like 'babelfish_conv_string_to_datetimeoffset';
                 proname                 | proparallel 
-----------------------------------------+-------------
 babelfish_conv_string_to_datetimeoffset | u
(1 row)
```

Anikait143 commented on test/JDBC/expected/datename_tzoffset-before-16_10-or-17_6-vu-prepare.out:
Since we are introducing these files first time why do we need to create before files for these?

forestkeeper replied:
Because this behavior is different from previous versions even we introduce it the first time, and we should add those into previous version upgrade schedule.

forestkeeper commented on test/JDBC/upgrade/17_5/schedule:
```diff
@@ -212,6 +212,7 @@ dateadd_internal_df
 datediff_big
 datediff
 dateadd
+datename_tzoffset-before-16_10-or-17_6
```
I think should add this into not just this schedule file, but also other schedule files as well.
